packages/daemon/src/service/pipeline.ts

```diff
@@ -48,6 +48,200 @@ function execAsync(cmd: string, opts: ExecOptions): Promise<string> {
+    const jobEvent = new JobStatusChangeEvent(
+      jobStatus,
+      step,
+      stepAction
+    );
+    this.opts.tracer.dispatch(jobEvent);
+  }
+  async runPlugin(type: PluginTypeI, ...args: any[]) {
+    const plugin = this.opts.plugins[type].plugin;
+    this.dispatchJobEvent(PipelineStatus.RUNNING, type, 'start');
+    const result = await this.opts.runnable.start(plugin, ...args);
+    this.dispatchJobEvent(PipelineStatus.RUNNING, type, 'end');
+    return result;
+  }
+
+  async runDataCollect(dataDir: string, modelPath: string): Promise<any> {
+    this.verifyPlugin('dataCollect');
+    // ensure the model dir exists
+    await fs.ensureDir(modelPath);
+    // run dataCollect to download dataset.
+    await this.runPlugin('dataCollect', this.getParams(this.opts.plugins.dataCollect.params, {
+      dataDir
+    }));
+  }
+  async runDataAccess(dataDir: string): Promise<any> {
+    this.verifyPlugin('dataAccess');
+    return this.runPlugin('dataAccess', this.getParams(this.opts.plugins.dataAccess.params, {
+      dataDir
+    }));
+  }
+
+  async runDatasetProcess(dataset: any): Promise<void> {
+    if (this.opts.plugins.datasetProcess) {
+      await this.runPlugin('datasetProcess', dataset, this.getParams(this.opts.plugins.datasetProcess.params));
+    }
+  }
+
+  async runDataProcess(dataset: any): Promise<void> {
+    if (this.opts.plugins.dataProcess) {
+      await this.runPlugin('dataProcess', dataset, this.getParams(this.opts.plugins.dataProcess.params));
+    }
+  }
+
+  async runModelDefine(dataset: any): Promise<ModelResult> {
+    return {
+      modelPlugin: this.opts.plugins.modelDefine.plugin,
+      model: await this.runPlugin('modelDefine', dataset, this.getParams(this.opts.plugins.modelDefine.params))
+    };
+  }
+
+  async runModelLoad(dataset: any, modelPath: string): Promise<ModelResult> {
+    return {
+      modelPlugin: this.opts.plugins.modelLoad.plugin,
+      model: await this.runPlugin('modelLoad', dataset, this.getParams(this.opts.plugins.modelLoad.params, {
+        // specify the recover path for model loader by default.
+        recoverPath: modelPath
+      }))
+    };
+  }
+
+  async runModelTrain(dataset: any, model: RunnableResponse, modelPath: string) {
+    return this.runPlugin('modelTrain', dataset, model, this.getParams(this.opts.plugins.modelTrain.params, {
+      modelPath
+    }));
+  }
+
+  async runModelEvaluate(dataset: any, model: RunnableResponse, modelPath: string): Promise<any> {
+    this.verifyPlugin('modelEvaluate');
+    return this.runPlugin('modelEvaluate', dataset, model, this.getParams(this.opts.plugins.modelEvaluate.params, {
+      modelDir: modelPath
+    }));
+  }
+  /**
+   * Generate the output package for a given job.
+   * @param job the job model for output.
+   * @param opts the options to used for generating the output.
+   */
+  async generateOutput(job: JobEntity, opts: GenerateOptions) {
```

Comment:
The JobRunner should not handle generating output IMHO.

Reply:
Yeah, it's better. Updated.